File: Concurrency/1116_Print_Zero_Even_Odd/Go/Solution.go
package main

import (
	"fmt"
	"sync"
)

// function type for passing function to methods
type printerFunc func(i int)

type ZeroEvenOdd struct {
	// number of iterations
	n int
	// mutex for printing zeros
	MtxZero *sync.Mutex
	// mutex for printing even numbers
	MtxEven *sync.Mutex
	// mutex for printing odd numbers
	MtxOdd *sync.Mutex
}

// zero printing method
func (zeo *ZeroEvenOdd) Zero(printNumber printerFunc) {
	for i := 0; i < zeo.n; i++ {
		// lock zero mutex
		zeo.MtxZero.Lock()

		// print zero
		printNumber(0)

		if i%2 == 0 {
			// unlock odd mutex if index is even
			zeo.MtxOdd.Unlock()
		} else {
			// unlock even mutex if index is odd
			zeo.MtxEven.Unlock()
		}
	}
}

// even printing method
func (zeo *ZeroEvenOdd) Even(printNumber printerFunc) {
	for i := 2; i < zeo.n; i += 2 {
		// lock even mutex
		zeo.MtxEven.Lock()
		// print index
		printNumber(i)
		// unlock zero mutex
		zeo.MtxZero.Unlock()
	}
}

// odd printing method
func (zeo *ZeroEvenOdd) Odd(printNumber printerFunc) {
	for i := 1; i < zeo.n; i += 2 {
		// lock odd mutex
		zeo.MtxOdd.Lock()
		// print index
		printNumber(i)
		// unlock zero mutex
		zeo.MtxZero.Unlock()
	}
}

func main() {
	// wait group to synchronize goroutines
	wg := sync.WaitGroup{}
	// add 3 wait groups
	wg.Add(3)

	// number of iterations
	n := 12
	// create mutexes
	mtxZero := sync.Mutex{}
	mtxEven := sync.Mutex{}
	mtxOdd := sync.Mutex{}
	// lock even & odd mutexes
	mtxEven.Lock()
	mtxOdd.Lock()

	// printer function
	printNumber := func(i int) {
		fmt.Println(i)
	}

	// create solution object
	zeo := ZeroEvenOdd{
		n:       n,
		MtxZero: &mtxZero,
		MtxEven: &mtxEven,
		MtxOdd:  &mtxOdd,
	}

	// zero printing goroutine
	go func() {
		zeo.Zero(printNumber)
		// signal wait group termination
		wg.Done()
	}()

	// even number printing goroutine
	go func() {
		zeo.Even(printNumber)
		// signal wait group termination
		wg.Done()
	}()

	// odd number printing goroutine
	go func() {
		zeo.Odd(printNumber)
		// signal wait group termination
		wg.Done()
	}()

	// wait for wait groups to terminate
	wg.Wait()
}

File: Concurrency/1116_Print_Zero_Even_Odd/Python/Solution.py
import threading

class ZeroEvenOdd:
    def __init__(self, n):
        self.n = n
        # mutex lock for zero()
        self.mutex_zero = threading.Lock()
        # mutex lock for even()
        self.mutex_even = threading.Lock()
        # mutex lock for odd()
        self.mutex_odd = threading.Lock()
        # lock mutex even
        self.mutex_even.acquire()
        # lock mutex odd
        self.mutex_odd.acquire()

    # printNumber(x) outputs "x", where x is an integer.
    def zero(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(self.n):
            # wait till mutex zero is unlocked, then lock it
            self.mutex_zero.acquire()
            # print zero
            printNumber(0)
            # if index is odd
            if i % 2 == 1:
                # unlock mutex even
                self.mutex_even.release()
            else:
                # unlock mutex odd
                self.mutex_odd.release()

        # if n is odd, unlock mutex even
        if self.n % 2 == 1:
            # unlock mutex even
            self.mutex_even.release()

    def even(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(2, self.n + 1, 2):
            # wait till mutex even is unlocked, then lock it
            self.mutex_even.acquire()
            # print even number
            printNumber(i)
            # unlock mutex zero
            self.mutex_zero.release()

    def odd(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(1, self.n + 1, 2):
            # wait till mutex odd is unlocked, then lock it
            self.mutex_odd.acquire()
            # print odd number
            printNumber(i)
            # unlock mutex zero
            self.mutex_zero.release()
